manifests/zone.pp

```diff
@@ -14,46 +15,59 @@
     $allow_transfer = [],
     $allow_query    = [],
     $also_notify    = [],
-    $zone           = $title,
-    $contact        = "root.${title}.",
+    $zone           = undef,
+    $contact        = undef,
     $zonefilepath   = $::dns::zonefilepath,
-    $filename       = "db.${title}",
+    $filename       = undef,
     $manage_file    = true,
     $forward        = 'first',
     $forwarders     = [],
     $dns_notify     = undef,
 ) {
 
   validate_bool($reverse, $manage_file)
-  validate_array($masters, $allow_transfer, $allow_query, $forwarders, $also_notify)
+  validate_array($masters, $allow_transfer, $allow_query, $forwarders, $also_notify, $target_views)
   validate_re($forward, '^(first|only)$', 'Only \'first\' or \'only\' are valid values for forward field')
   if $dns_notify {
     validate_re($dns_notify, '^(yes|no|explicit)$', 'Only \'yes\', \'no\', or \'explicit\' are valid values for dns_notify field')
   }
 
-  $zonefilename = "${zonefilepath}/${filename}"
+  if $zone     == undef { $_zone = $title } else { $_zone = $zone }
+  if $contact  == undef { $_contact  = "root.${_zone}." } else { $_contact = $contact }
+  if $filename == undef { $_filename = "db.${_zone}" } else { $_filename = $filename }
+
+  $zonefilename = "${zonefilepath}/${_filename}"
+
+  if $::dns::enable_views {
+    if $target_views == undef or size($target_views) == 0 {
+      warning('You seem to mix BIND views with global zones, which will probably fail')
+      $_target_views = ['_GLOBAL_']
+    } else {
+      $_target_views = $target_views
+    }
+  } else {
+      $_target_views = ['_GLOBAL_']
+  }
 
   if $zonetype == 'slave' {
     $_dns_notify = pick($dns_notify, 'no')
   } else {
     $_dns_notify = $dns_notify
   }
 
-  concat::fragment { "dns_zones+10_${zone}.dns":
-    target  => $::dns::publicviewpath,
-    content => template('dns/named.zone.erb'),
-    order   => "10-${zone}",
-  }
+  create_viewzones($_target_views)
 
   if $manage_file {
-    file { $zonefilename:
-      ensure  => file,
-      owner   => $dns::user,
-      group   => $dns::group,
-      mode    => '0644',
-      content => template('dns/zone.header.erb'),
-      replace => false,
-      notify  => Service[$::dns::namedservicename],
+    unless defined(File[$zonefilename]) {
```

Comment:
Hm, maybe we cannot drop it, or I need at least a hint considering the following use case:
- define a master dns::zone *master_zone*
- define custom content for the zonefile for *master_zone* outside of ::dns

To do that, I need to have `manage_file => true` because of
```
<% if @manage_file == true -%>
    file "<%= @zonefilename %>";
<% end -%>
```
If we drop the `unless defined(File[$zonefilename])` the content attribute is set in dns::zone, and I am not able to override it in my own Puppet code.